core/domain/user_jobs_one_off.py

```diff
@@ -208,75 +212,49 @@ def map(item):
                 if len(completed_nodes) != len(all_nodes):
                     learner_progress_services.mark_story_as_incomplete(
                         user_id, story.id)
+                else:
+                    learner_progress_services.mark_story_as_completed(
+                        user_id, story.id)
+
+                completed_story_ids = (
```

Comment:
Why don't you have this for loop just handle stories, and maintain a running list of newly-completed stories? Then, do the calculation for completed topics at the end, separately, rather than once per loop.

Here you have a for loop with lots of O(1) datastore queries in it and that will take time, it should be a get-multi thing.

Reply:
Done.